prow/gcsupload/run.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gcsupload
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/option"
+	"k8s.io/test-infra/prow/kube"
+	"k8s.io/test-infra/prow/pod-utils/downwardapi"
+	"k8s.io/test-infra/prow/pod-utils/gcs"
+)
+
+// Run will upload files to GCS as prescribed by
+// the options. Any extra files can be passed as
+// a parameter and will have the prefix prepended
+// to their destination in GCS, so the caller can
+// operate relative to the base of the GCS dir.
+func (o Options) Run(extra map[string]gcs.UploadFunc) error {
```

Comment:
Test coverage?

Reply:
Again would like to hear best approaches for testing, but will tackle in follow-up.

The call right now does:

 - create builder: can break this out into a method and test the `case` statement
 - `ResolveSpecFromEnv()`: one `os.LookupEnv` and one `json.Unmarshal` call, what to test?
 - construction of `gcsPath`: meat of this (`gcs.PathForSpec()`) is fully tested, only additional logic here is to append the `o.SubDir` which I guess we can also factor out into a `func` and test? Sacrificed readability to test one `if/else`
 - alias addition: `gcs.AliasForSpec()` is fully tested, only other logic here is to build the full path
 - upload latest builds: `gcs.LatestBuildForSpec()` is fully tested, only logic here is to iterate on them and add to the map
 - item selection: these are calls to `os.Stat` and `path.Walk` -- not super interested in fleshing out test directory trees in bash to test this, worried that would be more harm than good
 - when not dry run, construct client and upload: `gcs.Upload()` is tested, construction of a client doesn't seem worth testing

Reply:
So there are two cases here -- in the first case we could factor out one or two or three lines of code into a standalone function and unit-test that while sacrificing readability for the overall flow of the function; in the second case we're doing nothing but calling filesystem or environment APIs and it's not simple to mock those out.

Reply:
Integration test with filesystem muckery SGTM --> https://github.com/kubernetes/test-infra/issues/7532